api/src/test/java/com/spotify/ffwd/util/HighFrequencyDetectorTest.java

```diff
@@ -0,0 +1,182 @@
+                .toInstance(minNumberOfTriggers);
+            bind(HighFrequencyDetectorStatistics.class).toInstance(statistics);
+          }
+        });
+
+    final Injector injector = Guice.createInjector(modules);
+
+    return injector.getInstance(HighFrequencyDetector.class);
+  }
+
+  @Test
+  public void testHighFrequencyDetection() {
+    // verifies if high frequency metrics are captured and sortable
+    List<Metric> finalList = new ArrayList<>();
+    for (int y = 0; y < 3; y++) {
+      for (int x = 0; x < 3; x++) {
+        List<Metric> list = new ArrayList<>();
+        for (int i = 0; i < 30; i++) {
+          list.add(
+              new Metric(
+                  "KEY" + y,
+                  42.0 + i,
+                  new Date(),
+                  ImmutableSet.of(),
+                  Map.of("tag1", "value1", "what", "fun"),
+                  ImmutableMap.of(),
+                  null));
+        }
+
+        finalList.addAll(detector.detect(list));
+      }
+    }
+    assertEquals(90, finalList.size());
+  }
+
+  @Test
+  public void testLargeNumberOfHighFreqMetrics() {
+    // verifies if high frequency metrics are captured and sortable - up to 1k
+    List<Metric> finalList = new ArrayList<>();
+    for (int y = 0; y < 1000; y++) { // <-- number of high freq metrics kept in memory
+      for (int x = 0; x < 3; x++) { // <-- making sure it will kick in the detection
+        List<Metric> list = new ArrayList<>();
+        for (int i = 0; i < 20; i++) { // <--- reasonable batches
+          list.add(
+              new Metric(
+                  "KEY" + y,
+                  42.0 + i,
+                  new Date(),
+                  ImmutableSet.of(),
+                  Map.of("tag1", "value1", "what", "fun"),
+                  ImmutableMap.of(),
+                  null));
+        }
+
+        finalList.addAll(detector.detect(list));
+      }
+    }
+    assertEquals(20000, finalList.size());
+  }
+
+  @Test
+  public void testHighFreqStatsSwapping() {
+    // sleeps between detect() method call so due to
+    // highFrequencyDataRecycleMS set to 1 second for this test
+    // it will reset high freq stats and allow more metrics to be sent
+    List<Metric> finalList = new ArrayList<>();
+    for (int y = 0; y < 5; y++) { // <-- number of high freq metrics kept in memory
+      for (int x = 0; x < 3; x++) { // <-- making sure it will kick in the detection
+        List<Metric> list = new ArrayList<>();
+        for (int i = 0; i < 20; i++) { // <--- reasonable batches
+          list.add(
+              new Metric(
+                  "KEY" + y,
+                  42.0 + i,
+                  new Date(),
+                  ImmutableSet.of(),
+                  Map.of("tag1", "value1", "what", "fun"),
+                  ImmutableMap.of(),
+                  null));
+        }
+
+        finalList.addAll(shortSwapDetector.detect(list));
+        sleep(300);
```

Comment:
so this will sleep for a total of 300ms * 3 * 5 = 4.5 seconds. Is that cool w.r.t. making our unit tests run slower?

Reply:
We need to recreate scenario to force stats swapping and I think it is relatively quick. I've made all the parameters as small as possible.